llvm/lib/Transforms/Utils/SYCLModuleSplit.cpp

```diff
@@ -0,0 +1,935 @@
+#include "llvm/IR/Function.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/IR/PassManagerImpl.h"
+#include "llvm/IRPrinter/IRPrintingPasses.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/LineIterator.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Transforms/IPO.h"
+#include "llvm/Transforms/IPO/GlobalDCE.h"
+#include "llvm/Transforms/IPO/Internalize.h"
+#include "llvm/Transforms/IPO/StripDeadPrototypes.h"
+#include "llvm/Transforms/IPO/StripSymbols.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+#include "llvm/Transforms/Utils/SYCLUtils.h"
+
+#include <algorithm>
+#include <map>
+#include <utility>
+#include <variant>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "sycl_module_split"
+
+namespace {
+// Identifying name for global scope
+constexpr char GLOBAL_SCOPE_NAME[] = "<GLOBAL>";
+constexpr char SYCL_SCOPE_NAME[] = "<SYCL>";
+
+EntryPointsGroupScope selectDeviceCodeGroupScope(const Module &M,
+                                                 IRSplitMode Mode,
+                                                 bool AutoSplitIsGlobalScope) {
+  switch (Mode) {
+  case SPLIT_PER_TU:
+    return Scope_PerModule;
+
+  case SPLIT_PER_KERNEL:
+    return Scope_PerKernel;
+
+  case SPLIT_AUTO: {
+    if (AutoSplitIsGlobalScope)
+      return Scope_Global;
+
+    // At the moment, we assume that per-source split is the best way of
+    // splitting device code and can always be used except for cases handled
+    // above.
+    return Scope_PerModule;
+  }
+
+  case SPLIT_NONE:
+    return Scope_Global;
+  }
+
+  llvm_unreachable("unsupported split mode");
+}
+
+// Return true if the function is a SPIRV or SYCL builtin, e.g.
+// _Z28__spirv_GlobalInvocationId_xv
+bool isSpirvSyclBuiltin(StringRef FName) {
+  if (!FName.consume_front("_Z"))
+    return false;
+  // now skip the digits
+  FName = FName.drop_while([](char C) { return std::isdigit(C); });
+
+  return FName.starts_with("__spirv_") || FName.starts_with("__sycl_");
+}
+
+// Return true if the function name starts with "__builtin_"
+bool isGenericBuiltin(StringRef FName) {
+  return FName.starts_with("__builtin_");
+}
+
+bool isKernel(const Function &F) {
+  return F.getCallingConv() == CallingConv::SPIR_KERNEL ||
+         F.getCallingConv() == CallingConv::AMDGPU_KERNEL;
+}
+
+bool isEntryPoint(const Function &F, bool EmitOnlyKernelsAsEntryPoints) {
+  // Skip declarations, if any: they should not be included into a vector of
+  // entry points groups or otherwise we will end up with incorrectly generated
+  // list of symbols.
+  if (F.isDeclaration())
+    return false;
+
+  // Kernels are always considered to be entry points
+  if (isKernel(F))
+    return true;
+
+  if (!EmitOnlyKernelsAsEntryPoints) {
```

Comment:
Can we consider just enabling EmitOnlyKernelsAsEntryPoints by default?

Reply:
Removed the option.